com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java

```diff
@@ -1295,22 +1295,27 @@ public void removeNode(final ZigBeeNode node) {
             networkNodes.remove(node.getIeeeAddress());
         }
 
-        synchronized (this) {
-            for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
-                NotificationService.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        listener.nodeRemoved(node);
-                    }
-                });
-            }
+        for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
+            NotificationService.execute(new Runnable() {
+                @Override
+                public void run() {
+                    listener.nodeRemoved(node);
+                }
+            });
+        }
 
-            node.shutdown();
+        node.shutdown();
 
-            if (networkStateSerializer != null) {
-                networkStateSerializer.serialize(this);
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (ZigBeeNetworkManager.this) {
+                    if (networkStateSerializer != null) {
```

Comment:
Wouldn't it be better to put this check outside of the thread?

Reply:
If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).

Reply:
If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).

Reply:
If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).